stan/math/rev/functor/gradient.hpp

```diff
@@ -51,6 +55,67 @@ void gradient(const F& f, const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
   grad_fx = x_var.adj();
 }
 
+/**
+ * Calculate the value and the gradient of the specified function
+ * at the specified argument.
+ *
+ * <p>The functor must implement
+ *
+ * <code>
+ * var
+ * operator()(const
+ * Eigen::Matrix<var, Eigen::Dynamic, 1>&)
+ * </code>
+ *
+ * using only operations that are defined for
+ * <code>var</code>.  This latter constraint usually
+ * requires the functions to be defined in terms of the libraries
+ * defined in Stan or in terms of functions with appropriately
+ * general namespace imports that eventually depend on functions
+ * defined in Stan.
+ *
+ * The evaluated gradient is stored into the object whose data
+ * begins at <code>*first_grad_fx</code> and ends at
+ * <code>*last_grad_fx</code>.  The caller is responsible for
+ * ensuring the size of the object pointed to by
+ * <code>first_grad_fx</code> matches the size of the argument
+ * <code>x</code>.
+ *
+ * <p>Time and memory usage is on the order of the size of the
+ * fully unfolded expression for the function applied to the
+ * argument, independently of dimension.
+ *
+ * @tparam F Type of function
+ * @param[in] f Function
+ * @param[in] x Argument to function
+ * @param[out] fx Function applied to argument
+ * @param[out] grad_fx Gradient of function at argument
+ * @throw std::invalid_argument if the iterator isn't the right size
+ * to hold the gradients
+ */
+template <typename F, typename InputIt>
+void gradient(const F& f, const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
+              double& fx, InputIt first_grad_fx, InputIt last_grad_fx) {
+  nested_rev_autodiff nested;
+
+  if (last_grad_fx - first_grad_fx != x.size()) {
+    std::stringstream s;
+    s << "gradient(): iterator and gradient different sizes; iterator size = "
+      << last_grad_fx - first_grad_fx << "; grad size = " << x.size()
+      << std::endl;
+    throw std::invalid_argument(s.str());
+  }
+
+  Eigen::Matrix<var, Eigen::Dynamic, 1> x_var(x);
+  var fx_var = f(x_var);
+  fx = fx_var.val();
+  grad(fx_var.vi_);
+  for (Eigen::Matrix<var, Eigen::Dynamic, -1>::Index i = 0; i < x_var.size();
+       ++i) {
+    *first_grad_fx++ = x_var(i).adj();
```

Comment:
This is nbd / optional but using `.coeff()` will avoid a out of bounds check you are already doing manually in the above

```suggestion
    *first_grad_fx++ = x_var.coeff(i).adj();
```